mmv1/third_party/tgc_next/pkg/tfplan2cai/resolvers/advanced_resolver.go

```diff
@@ -0,0 +1,139 @@
+package resolvers
+
+import (
+	"slices"
+	"sort"
+	"strings"
+
+	"go.uber.org/zap"
+
+	"github.com/GoogleCloudPlatform/terraform-google-conversion/v7/pkg/tfplan2cai/models"
+	"github.com/GoogleCloudPlatform/terraform-google-conversion/v7/pkg/tfplan2cai/tfplan"
+
+	provider "github.com/GoogleCloudPlatform/terraform-google-conversion/v7/pkg/provider"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+// List of keyword to filter out in order to find the iam resource parent
+var filterList = []string{"etag", "id", "role", "members", "condition", "member"}
+
+type AdvancedPreResolver struct {
+	schema *schema.Provider
+
+	// For logging error / status information that doesn't warrant an outright failure
+	errorLogger *zap.Logger
+}
+
+func NewAdvancedResolver(errorLogger *zap.Logger) *AdvancedPreResolver {
+	return &AdvancedPreResolver{
+		schema:      provider.Provider(),
+		errorLogger: errorLogger,
+	}
+}
+
+func (r *AdvancedPreResolver) Resolve(jsonPlan []byte, resourceDataMap map[string][]*models.FakeResourceDataWithMeta) map[string][]*models.FakeResourceDataWithMeta {
+	// ReadChanges
+	planChanges, err := tfplan.ReadResourceChanges(jsonPlan)
+	if err != nil {
+		return resourceDataMap
+	}
+
+	// Read elements from the resouce config
+	resourceConfig, err := tfplan.ReadResourceConfigurations(jsonPlan)
+	if err != nil {
+		return resourceDataMap
+	}
+
+	// Key is iam resource, values are ids
+	idToAddressMap := make(map[string][]string)
+
+	for _, rc := range planChanges {
+		// Silently skip non-google resources
+		if !strings.HasPrefix(rc.Type, "google_") {
+			continue
+		}
+		// Handle iam resources, build an id for each of them and group them together
+		if strings.Contains(rc.Type, "iam") {
+			var keys []string
+			// Take all keys from Change.After and store them in a list
+			afterMap, ok := rc.Change.After.(map[string]interface{})
+			if ok {
+				for k := range afterMap {
+					if !slices.Contains(filterList, k) {
+						keys = append(keys, k)
+					}
+				}
+			}
+
+			// Take all keys from Change.AfterUnknown and store them in a list
+			afterUnknownMap, ok := rc.Change.AfterUnknown.(map[string]interface{})
+			if ok {
+				for k := range afterUnknownMap {
+					if !slices.Contains(filterList, k) {
+						keys = append(keys, k)
+					}
+				}
+			}
+
+			sort.Strings(keys)
+			resourceId := ""
+			// Build the id for each iam resource
+			for _, key := range keys {
+				// variable refers to the parent argument,
+
+				value, ok := afterMap[key]
+				if ok {
+					resourceId = resourceId + key + "/"
```

Comment:
This line can be combined with line 88. Only the value is a string, then add the key to the resource id.